docs/design/password-key-derivation.md

```diff
@@ -0,0 +1,66 @@
+
+This design doc describes the way cryptographic encryption keys will be derived from user provided passwords in a way that resists offline attacks and low entropy passwords.
+
+## Background
+
+A unique encryption key exists for every path inside of a bucket, and they are designed in a way that allows deriving a key from a parent and some path component. For example, if you had the key for `sj://bucket/path1/path2` then you could derive the key for `sj://bucket/path1/path2/path3`.
+
+That leaves open the question for how a root key is derived. Some requirements on that root key derivation include
+
+1. It's based on the user's choice (for example, a password).
+2. If the user inputs the same password on different machines, the same root key should be derived.
+
+These requirements allow users to be in full control of their encryption, and don't require users safely transporting high entropy (hard to remember) secrets to bootstrap new uplinks.
+
+This design accomodates more requirements that allow for additional features:
+
+3. A root key can be derived for any path in a bucket.
+4. Buckets root keys should not depend on the name of the bucket.
+5. A table of keys for low entropy passwords should not be possible. In other words, an attacker with knowledge of the algorithm should not be able to use a dictionary of common passwords and pre-compute what keys to check in the event of a data breach.
+6. There is a default that allows deriving keys for buckets when no bucket specific secret is provided.
+
+The third requirement allows having multiple encrypted domains to exist within a single bucket, allowing delegation of encryption. The fourth requirement allows renaming buckets without having to re-encrypt all of the paths inside of a bucket. The fifth requirement enhances security if a satellite breach happens. This is usually accomplished with [salting](https://en.wikipedia.org/wiki/Salt_(cryptography)). The sixth requirement allows providing a finite amount of information to a third party that allows them access to the keys for any bucket created in the future.
+
+## Design
+
+A default password can be chosen. There is no extra entropy or information available, so it will be run through [Argon2](https://en.wikipedia.org/wiki/Argon2) with a very high work load and a fixed salt. The output of that will be stored in a secure location. The higher the workload, the better, as it increases the cost for an offline attack, and only the output is used in subsequent steps, so it only needs to happen once.
+
+Satellites will add a new field to bucket metadata called `entropy` that just contains 32 random, stable bytes.
+
+The algorithm used to derive a root key for some path in some bucket for some secret is
+
+```
+bucketEntropy = getBucketMetadata(bucketName).entropy
+bucketRootKey = argon2id(salt:hmac(hash:sha256, secret:secret, data:bucketEntropy), secret:secret)
+pathRootKey   = hmac(hash:sha256, secret:bucketRootKey, data:encryptedPath)
```

Comment:
Oh, wait, what is `encryptedPath` in this context, if we are deriving s<sub>0</sub>?

Or does the below paragraph mean that `bucketRootKey` is s<sub>0</sub>, and `pathRootKey` is just being given as an example path derivation step for an encryptedPath that has exactly one path component?

Reply:
Nope, this is for a use case where, for example, someone has a bucket, and they want to allow users to have their own paths in that bucket, and the want the users to have their own encryption keys.

In that case, the s<sub>0</sub> exists at, maybe, `sj://bucket/user1`.

Reply:
Ah, I hadn't thought of that need. Clever.